### ISSUE TITLE
Disassembler DisassembleMethod fails with `Object reference not set to an instance of an object.`,

### ISSUE THREAD

**Issue by nietras** [dotnet/BenchmarkDotNet#612]
Running a benchmark with disassembler turned on for something like:
```
 [DisassemblyDiagnoser(printAsm: true, printSource: true, recursiveDepth: 7)]
    [SimpleJob(RunStrategy.Monitoring, launchCount: 1, warmupCount: 2, targetCount: 11)]
    //[RyuJitX64Job()]
    public class RandomSort
    {
        const int Length = 2000000;
        static readonly int[] _random = CreateRandomArray<int>(Length, i => i);
        int[] _work = new int[Length];

        [IterationSetup]
        public void IterationSetup()
        {
            //Console.WriteLine(nameof(IterationSetup));
            Array.Copy(_random, _work, Length);
        }

        [Benchmark(Baseline = true)]
        public void ArraySort()
        {
            // CLR code
            Array.Sort(_work);
        }

        [Benchmark]
        public void SpanSort()
        {
            new Span<int>(_work).Sort();
        }
}
```

after settings `recursiveDepth: 7` this fails with:

```
Failed to disassemble with following exception:
Object reference not set to an instance of an object.
   at Microsoft.Diagnostics.RuntimeExt.ClrSourceExtensions.FindNearestLine(PdbFunction function, Int32 ilOffset)
   at BenchmarkDotNet.Disassembler.Program.<GetSource>d__9.MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at BenchmarkDotNet.Disassembler.Program.DisassembleMethod(MethodInfo methodInfo, State state, Settings settings)
   at BenchmarkDotNet.Disassembler.Program.Disassemble(Settings settings, ClrRuntime runtime, State state)
   at BenchmarkDotNet.Disassembler.Program.Handle(Settings settings)
   at BenchmarkDotNet.Disassembler.Program.Main(String[] args)
```

**Comment by nietras** [comment]
cc: @adamsitnik

**Comment by nietras** [comment]
On that note is there any way to get dissassembly for runtime code? I.e. code I do not have the source for?

**Comment by adamsitnik** [comment]
@nietras the workaround: show only asm, no source code:

```cs
 [DisassemblyDiagnoser(printAsm: true, printSource: false, recursiveDepth: 7)]
```

Could you tell me which .NET version are you using? I need to reproduce this bug locally, the workaround is only temporal ofc.

**Comment by nietras** [comment]
Follow up, it works for `recursiveDepth: 5` a report is attached below. This does give some assembly for the `Array.Sort` but I need deeper depth I think.

[DotNetCross.Sorting.Benchmarks.RandomSort-disassembly-report.html.zip](https://github.com/dotnet/BenchmarkDotNet/files/1595889/DotNetCross.Sorting.Benchmarks.RandomSort-disassembly-report.html.zip)

**Comment by nietras** [comment]
@adamsitnik you can get the entire source for this from https://github.com/DotNetCross/Sorting/tree/span-sort that is `span-sort` branch of that repo. It is part of my work implementing Sort for spans in corefx. Any help appreciated ;) It is still quite early though...

**Comment by adamsitnik** [comment]
@nietras The exception you are getting is thrown when we try to get the C# code. If you set `printSource: false` it should work for any `recursiveDepth`.

**Comment by nietras** [comment]
@adamsitnik thanks for the reply. Alright, it is weird that it works for `recursiveDepth: 6` but not `recursiveDepth: 7`... not sure what changes. It would be great if you could actually `printSource: true` together with "runtime"/corefx code, of course. It would, of course, be even greater if the reference source for the "runtime"/corefx could be used and downloaded etc. But that is a lofty wish ;)

Also, it would be good if one could somehow limit the recursion so as to some basic type stuff does not get pulled in as it seems to do here, meaning the runtime code disassembly could avoid some types.

Also, (so many wishes :)), often method names etc get really long 
```
00007ffb`099617c0 System.SpanHelpers+SpanSortHelper`2[[System.Int32, System.Private.CoreLib],[System.SpanHelpers+ComparableComparer`1[[System.Int32, System.Private.CoreLib]], DotNetCross.Sorting]].PickPivotAndPartition(Int32 ByRef, Int32, Int32, ComparableComparer`1 ByRef)
```
meaning one has to scroll horizontally to see the method name. Perhaps this line could be split into multiple lines depending on its length being above some threshold... e.g. something like:
```
00007ffb`099617c0 System.SpanHelpers+SpanSortHelper`2
    [[System.Int32, System.Private.CoreLib],[System.SpanHelpers+ComparableComparer`1[[System.Int32, System.Private.CoreLib]], DotNetCross.Sorting]]
    .PickPivotAndPartition(Int32 ByRef, Int32, Int32, ComparableComparer`1 ByRef)
```
Perhaps, even doing some basic cleanup of base types e.g. replace `System.Int32` with `int` etc.

**Comment by nietras** [comment]
I'm having another issue, perhaps related to this. After adding another Class Library I am getting:
```
// AfterAll
Failed to disassemble with following exception:
Sequence contains no matching element
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   at BenchmarkDotNet.Disassembler.Program.Disassemble(Settings settings, ClrRuntime runtime, State state)
   at BenchmarkDotNet.Disassembler.Program.Handle(Settings settings)
   at BenchmarkDotNet.Disassembler.Program.Main(String[] args)

Unhandled Exception: System.InvalidOperationException: There is an error in XML document (0, 0). ---> System.Xml.XmlException: Root element is missing.
   at System.Xml.XmlTextReaderImpl.Throw(Exception e)
   at System.Xml.XmlTextReaderImpl.ParseDocumentContent()
   at System.Xml.XmlTextReaderImpl.Read()
   at System.Xml.XmlReader.MoveToContent()
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderDisassemblyResult.Read9_DisassemblyResult()
   --- End of inner exception stack trace ---
   at System.Xml.Serialization.XmlSerializer.Deserialize(XmlReader xmlReader, String encodingStyle, XmlDeserializationEvents events)
   at BenchmarkDotNet.Diagnosers.WindowsDisassembler.Dissasemble(DiagnoserActionParameters parameters)
   at BenchmarkDotNet.Diagnosers.DisassemblyDiagnoser.Handle(HostSignal signal, DiagnoserActionParameters parameters)
   at BenchmarkDotNet.Extensions.CommonExtensions.ForEach[T](IList`1 source, Action`1 command)
   at BenchmarkDotNet.Loggers.SynchronousProcessOutputLoggerWithDiagnoser.ProcessInput()
   at BenchmarkDotNet.Toolchains.DotNetCli.DotNetCliExecutor.Execute(Benchmark benchmark, ILogger logger, ArtifactsPaths artifactsPaths, IDiagnoser diagnoser, String executableName, IConfig config, IResolver resolver)
   at BenchmarkDotNet.Toolchains.DotNetCli.DotNetCliExecutor.Execute(ExecuteParameters executeParameters)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.Execute(ILogger logger, Benchmark benchmark, IToolchain toolchain, BuildResult buildResult, IConfig config, IResolver resolver)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.RunCore(Benchmark benchmark, ILogger logger, ReadOnlyConfig config, String rootArtifactsFolderPath, Func`2 toolchainProvider, IResolver resolver, BuildResult buildResult)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.Run(BenchmarkRunInfo benchmarkRunInfo, ILogger logger, String title, String rootArtifactsFolderPath, Func`2 toolchainProvider, IResolver resolver, List`1 artifactsToCleanup)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.Run(BenchmarkRunInfo benchmarkRunInfo, Func`2 toolchainProvider)
   at DotNetCross.Sorting.Benchmarks.Program.Main(String[] args)
```
in the same repo for commit `4ee56c93f7c506453f8c6034859d56779cc8df2b`. That is the `span-sort` branch. All I did was add a new class library and some new benchmark code `NewRandomSort` based on this.

**Comment by nietras** [comment]
I also have some warnings during built if targetting benchmark project to `netcoreapp2.0`:
```
E:\oss\DotNetCross\Sorting\tests\DotNetCross.Sorting.Benchmarks\DotNetCross.Sorting.Benchmarks.csproj : warning NU1701: Package 'BenchmarkDotNet.Diagnostics.Windows 0.10.11.374' was restored using '.NETFramework,Version=v4.6.1' instead of the project target framework '.NETCoreApp,Version=v2.0'. This package may not be fully compatible with your project. [E:\oss\DotNetCross\Sorting\Sorting.sln]
E:\oss\DotNetCross\Sorting\tests\DotNetCross.Sorting.Benchmarks\DotNetCross.Sorting.Benchmarks.csproj : warning NU1701: Package 'BenchmarkDotNet.Diagnostics.Windows 0.10.11.374' was restored using '.NETFramework,Version=v4.6.1' instead of the project target framework '.NETCoreApp,Version=v2.0'. This package may not be fully compatible with your project.
    2 Warning(s)
```
but if I target `net471` I get issues with running since it then tries to built with `netcoreapp1.1`...

**Comment by nietras** [comment]
The `Sequence contains no matching element` issue is a big issue for me. It seems to occur as soon as the recursion goes into runtime code or something. Even with just `[DisassemblyDiagnoser(printAsm: true, recursiveDepth: 2)]` it is thrown now when using say `IntPtr` in the code. That is even without `printSource: true`.

This means I can't use the disassembly diagnoser for the code I am working on... ☹️

**Comment by nietras** [comment]
The following simple benchmark code will cause the Exception:
```csharp
    public static class IntPtrHelper
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public unsafe static IntPtr Multiply(this IntPtr a, int factor)
        {
            return (sizeof(IntPtr) == sizeof(int))
                ? new IntPtr((int)a * factor)
                : new IntPtr((long)a * factor);
        }
    }

    [DisassemblyDiagnoser(recursiveDepth: 2)]
    [SimpleJob(RunStrategy.Monitoring, launchCount: 1, warmupCount: 2, targetCount: 11)]
    public class IntPtrHelperBenchmark
    {
        [Benchmark]
        public IntPtr Ctor()
        {
            return new IntPtr(42).Multiply(4);
        }
    }
```
With version `0.10.11.374` on Windows the console window has this in it, the `SingleOrDefault` assumption seems to be wrong here.
```
// ***** BenchmarkRunner: Start   *****
// Found benchmarks:
//   IntPtrHelperBenchmark.Ctor: Job-HQUNAP(LaunchCount=1, RunStrategy=Monitoring, TargetCount=11, WarmupCount=2)

// Validating benchmarks:
// ***** Building 1 benchmark(s) in Parallel: Start   *****
// ***** Done, took 00:00:05 (5.37 sec)   *****
// **************************
// Benchmark: IntPtrHelperBenchmark.Ctor: Job-HQUNAP(LaunchCount=1, RunStrategy=Monitoring, TargetCount=11, WarmupCount=2)
// *** Execute ***
// Launch: 1 / 1
// BeforeAnythingElse

// Benchmark Process Environment Information:
// Runtime=.NET Core 2.0.3 (Framework 4.6.25815.02), 64bit RyuJIT
// GC=Concurrent Workstation
// Job: Job-RCEJDB(LaunchCount=1, RunStrategy=Monitoring, TargetCount=11, WarmupCount=2)

MainWarmup  1: 1 op, 671186.55 ns, 671.1865 us/op
MainWarmup  2: 1 op, 2792.73 ns, 2.7927 us/op

// BeforeMainRun
MainTarget  1: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget  2: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget  3: 1 op, 310.30 ns, 310.3035 ns/op
MainTarget  4: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget  5: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget  6: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget  7: 1 op, 310.30 ns, 310.3035 ns/op
MainTarget  8: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget  9: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget 10: 1 op, 0.00 ns, 0.0000 ns/op
MainTarget 11: 1 op, 0.00 ns, 0.0000 ns/op

// AfterMainRun
Result  1: 1 op, 0.00 ns, 0.0000 ns/op
Result  2: 1 op, 0.00 ns, 0.0000 ns/op
Result  3: 1 op, 310.30 ns, 310.3035 ns/op
Result  4: 1 op, 0.00 ns, 0.0000 ns/op
Result  5: 1 op, 0.00 ns, 0.0000 ns/op
Result  6: 1 op, 0.00 ns, 0.0000 ns/op
Result  7: 1 op, 310.30 ns, 310.3035 ns/op
Result  8: 1 op, 0.00 ns, 0.0000 ns/op
Result  9: 1 op, 0.00 ns, 0.0000 ns/op
Result 10: 1 op, 0.00 ns, 0.0000 ns/op
Result 11: 1 op, 0.00 ns, 0.0000 ns/op
GC:  0 0 0 0 0

// AfterAll
Failed to disassemble with following exception:
Sequence contains more than one matching element
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at BenchmarkDotNet.Disassembler.Program.GetMethod(MethodReference methodReference, ClrType declaringType)
   at BenchmarkDotNet.Disassembler.Program.EnqueueAllCalls(State state, ICollection`1 ilInstructions, Int32 depth)
   at BenchmarkDotNet.Disassembler.Program.DisassembleMethod(MethodInfo methodInfo, State state, Settings settings)
   at BenchmarkDotNet.Disassembler.Program.Disassemble(Settings settings, ClrRuntime runtime, State state)
   at BenchmarkDotNet.Disassembler.Program.Handle(Settings settings)
   at BenchmarkDotNet.Disassembler.Program.Main(String[] args)

Unhandled Exception: System.InvalidOperationException: There is an error in XML document (0, 0). ---> System.Xml.XmlException: Root element is missing.
   at System.Xml.XmlTextReaderImpl.Throw(Exception e)
   at System.Xml.XmlTextReaderImpl.ParseDocumentContent()
   at System.Xml.XmlTextReaderImpl.Read()
   at System.Xml.XmlReader.MoveToContent()
   at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReaderDisassemblyResult.Read9_DisassemblyResult()
   --- End of inner exception stack trace ---
   at System.Xml.Serialization.XmlSerializer.Deserialize(XmlReader xmlReader, String encodingStyle, XmlDeserializationEvents events)
   at BenchmarkDotNet.Diagnosers.WindowsDisassembler.Dissasemble(DiagnoserActionParameters parameters)
   at BenchmarkDotNet.Diagnosers.DisassemblyDiagnoser.Handle(HostSignal signal, DiagnoserActionParameters parameters)
   at BenchmarkDotNet.Extensions.CommonExtensions.ForEach[T](IList`1 source, Action`1 command)
   at BenchmarkDotNet.Loggers.SynchronousProcessOutputLoggerWithDiagnoser.ProcessInput()
   at BenchmarkDotNet.Toolchains.DotNetCli.DotNetCliExecutor.Execute(Benchmark benchmark, ILogger logger, ArtifactsPaths artifactsPaths, IDiagnoser diagnoser, String executableName, IConfig config, IResolver resolver)
   at BenchmarkDotNet.Toolchains.DotNetCli.DotNetCliExecutor.Execute(ExecuteParameters executeParameters)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.Execute(ILogger logger, Benchmark benchmark, IToolchain toolchain, BuildResult buildResult, IConfig config, IResolver resolver)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.RunCore(Benchmark benchmark, ILogger logger, ReadOnlyConfig config, String rootArtifactsFolderPath, Func`2 toolchainProvider, IResolver resolver, BuildResult buildResult)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.Run(BenchmarkRunInfo benchmarkRunInfo, ILogger logger, String title, String rootArtifactsFolderPath, Func`2 toolchainProvider, IResolver resolver, List`1 artifactsToCleanup)
   at BenchmarkDotNet.Running.BenchmarkRunnerCore.Run(BenchmarkRunInfo benchmarkRunInfo, Func`2 toolchainProvider)
   at DotNetCross.Sorting.Benchmarks.Program.Main(String[] args)
// * Artifacts cleanup *
Press any key to continue . . .
```

**Comment by nietras** [comment]
The offending code seems to be:
```csharp
        static ClrMethod GetMethod(MethodReference methodReference, ClrType declaringType)
        {
            var methodsWithSameToken = declaringType.Methods
                .Where(method => method.MetadataToken == methodReference.MetadataToken.ToUInt32()).ToArray();

            if (methodsWithSameToken.Length == 1) // the most common case
                return methodsWithSameToken[0];
            if (methodsWithSameToken.Length > 1 && methodReference.MetadataToken.ToUInt32() != default(UInt32)) 
                // usually one is NGened, the other one is not compiled (looks like a ClrMD bug to me)
                return methodsWithSameToken.Single(method => method.CompilationType != MethodCompilationType.None);

            // comparing metadata tokens does not work correctly for some NGENed types like Random.Next, System.Threading.Monitor & more
            // Mono.Cecil reports different metadata token value than ClrMD for the same method 
            // so the last chance is to try to match them by... name (I don't like it, but I have no better idea for now)
            var unifiedSignature = CecilNameToClrmdName(methodReference);

            return declaringType
                .Methods
                .SingleOrDefault(method => method.GetFullSignature() == unifiedSignature);
        }
```
Specifically:
```csharp
            // comparing metadata tokens does not work correctly for some NGENed types like Random.Next, System.Threading.Monitor & more
            // Mono.Cecil reports different metadata token value than ClrMD for the same method 
            // so the last chance is to try to match them by... name (I don't like it, but I have no better idea for now)
            var unifiedSignature = CecilNameToClrmdName(methodReference);

            return declaringType
                .Methods
                .SingleOrDefault(method => method.GetFullSignature() == unifiedSignature);
```
Added a PR for this in https://github.com/dotnet/BenchmarkDotNet/pull/615

**Comment by adamsitnik** [comment]
@nietras I was able to reporoduce your problem. I have pushed a fix for it.

Once [this](https://ci.appveyor.com/project/dotnetfoundation/benchmarkdotnet/build/0.10.11.376) build is done (35 minutes from now) you should be able to test it. Please let me know if it solved the issue.

I have also described how to debug the disassembler in our [docs](https://github.com/dotnet/BenchmarkDotNet/blob/7038155d914e5679696b17d18524e8066256d14e/docs/guide/Contributing/Disassembler.md) in case you would like to contribute more to it ;)

**Comment by nietras** [comment]
@adamsitnik thanks! Nice with the docs 👍

**Comment by xoofx** [comment]
Ah! I had also this bug `Unhandled Exception: System.InvalidOperationException: There is an error in XML document (0, 0). ---> System.Xml.XmlException: Root element is missing.`... but with 0.10.12, is it fixed there or only on latest master?

**Comment by adamsitnik** [comment]
Hello @xoofx 

This particular fix is part of 0.10.12. So if you hit a problem with 0.10.12 please create a new issue with repro case.
